src/swap/alice.rs

```diff
@@ -0,0 +1,262 @@
+//! Alice's perspective of the swap.
+//!
+//! In Nectar we always take the role of Bob in the swap, so our local
+//! representation of the other party, Alice, is a component that
+//! watches the two blockchains involved in the swap.
+
+use crate::swap::{
+    ethereum::{self, ethereum_latest_time},
+    SafeToFund, SafeToRedeem, {hbit, herc20},
+};
+use chrono::NaiveDateTime;
+use comit::{
+    btsieve::{ethereum::ReceiptByHash, BlockByHash, LatestBlock},
+    SecretHash, Timestamp,
+};
+use std::sync::Arc;
+
+#[derive(Clone, Debug)]
+pub struct WatchOnlyAlice<AC, BC> {
+    pub alpha_connector: Arc<AC>,
+    pub beta_connector: Arc<BC>,
+    pub secret_hash: SecretHash,
+    pub start_of_swap: NaiveDateTime,
+}
+
+#[async_trait::async_trait]
+impl<AC, BC> hbit::Fund for WatchOnlyAlice<AC, BC>
+where
+    AC: LatestBlock<Block = bitcoin::Block>
+        + BlockByHash<Block = bitcoin::Block, BlockHash = bitcoin::BlockHash>,
+    BC: Send + Sync,
+{
+    async fn fund(&self, params: &hbit::Params) -> anyhow::Result<hbit::CorrectlyFunded> {
+        let event =
+            hbit::watch_for_funded(self.alpha_connector.as_ref(), &params, self.start_of_swap)
+                .await?;
+
+        Ok(event)
+    }
+}
+
+#[async_trait::async_trait]
+impl<AC, BC> herc20::RedeemAsAlice for WatchOnlyAlice<AC, BC>
+where
+    AC: Send + Sync,
+    BC: LatestBlock<Block = ethereum::Block>
+        + BlockByHash<Block = ethereum::Block, BlockHash = ethereum::Hash>
+        + ReceiptByHash,
+{
+    async fn redeem(
+        &self,
+        _params: &herc20::Params,
+        deploy_event: herc20::Deployed,
+    ) -> anyhow::Result<herc20::Redeemed> {
+        let event = herc20::watch_for_redeemed(
+            self.beta_connector.as_ref(),
+            self.start_of_swap,
+            deploy_event,
+        )
+        .await?;
+
+        Ok(event)
+    }
+}
+
+#[async_trait::async_trait]
+impl<AC, BC> hbit::Refund for WatchOnlyAlice<AC, BC>
+where
+    AC: LatestBlock<Block = bitcoin::Block>
+        + BlockByHash<Block = bitcoin::Block, BlockHash = bitcoin::BlockHash>,
+    BC: Send + Sync,
+{
+    async fn refund(
+        &self,
+        params: &hbit::Params,
+        fund_event: hbit::CorrectlyFunded,
+    ) -> anyhow::Result<hbit::Refunded> {
+        let event = hbit::watch_for_refunded(
+            self.alpha_connector.as_ref(),
+            params,
+            fund_event.location,
+            self.start_of_swap,
+        )
+        .await?;
+
+        Ok(event)
+    }
+}
+
+#[async_trait::async_trait]
+impl<AC, BC> SafeToFund for WatchOnlyAlice<AC, BC>
+where
+    BC: LatestBlock<Block = ethereum::Block>,
+    AC: Send + Sync,
+{
+    async fn is_safe_to_fund(&self, beta_expiry: Timestamp) -> anyhow::Result<bool> {
+        let ethereum_time = ethereum_latest_time(self.beta_connector.as_ref()).await?;
+        // TODO: Apply a buffer depending on the blocktime and how
```

Comment:
I'd rather remove the todo and rename the function at this stage. The "buffer" should be a properly discussed subject out of scope of the nectar project because it's not that straightforward.
I actually had a shaped idea around that, need to discuss with @bonomat.

Reply:
I tend to agree with you. I'll remove the TODOs.